packages/playwright-core/src/server/browserContext.ts

```diff
@@ -610,11 +679,56 @@ export abstract class BrowserContext extends SdkObject {
         for (const originState of state.origins) {
           const frame = page.mainFrame();
           await frame.goto(metadata, originState.origin);
-          await frame.evaluateExpression(`
-            originState => {
-              for (const { name, value } of (originState.localStorage || []))
-                localStorage.setItem(name, value);
-            }`, { isFunction: true, world: 'utility' }, originState);
+
+          for (const dbInfo of (originState.indexedDB || [])) {
+            for (const store of dbInfo.stores) {
+              for (const record of store.records) {
+                if (record.key !== undefined)
+                  record.key = utilitySerializers.parseEvaluationResultValue(JSON.parse(record.key));
+                record.value = utilitySerializers.parseEvaluationResultValue(JSON.parse(record.value));
+              }
+            }
+          }
+
+          async function _restoreStorageState(originState: channels.OriginStorage) {
+            for (const { name, value } of (originState.localStorage || []))
+              localStorage.setItem(name, value);
+
+            function idbRequestToPromise<T extends IDBOpenDBRequest | IDBRequest>(request: T) {
+              return new Promise<T['result']>((resolve, reject) => {
+                request.addEventListener('success', () => resolve(request.result));
+                request.addEventListener('error', () => reject(request.error));
+              });
+            }
+
+            await Promise.all((originState.indexedDB ?? []).map(async dbInfo => {
+              const openRequest = indexedDB.open(dbInfo.name, dbInfo.version);
+              openRequest.addEventListener('upgradeneeded', () => {
+                const db = openRequest.result;
+                for (const store of dbInfo.stores) {
+                  const objectStore = db.createObjectStore(store.name, { autoIncrement: store.autoIncrement, keyPath: store.keyPathArray ?? store.keyPath });
+                  for (const index of store.indexes)
+                    objectStore.createIndex(index.name, index.keyPathArray ?? index.keyPath!, { unique: index.unique, multiEntry: index.multiEntry });
+                }
+              });
+
+              const db = await idbRequestToPromise(openRequest);
```

Comment:
So, this will end up in `success` after the `upgradeneeded` event handler has run? I cannot find this information on MDN, unfortunately.